pkg/server/handler/sso_customtoken.go

```diff
@@ -0,0 +1,294 @@
+	keyFunc     jwt.Keyfunc
+	TokenString string               `mapstructure:"token"`
+	Token       *jwt.Token           `mapstructure:"-"`
+	Claims      ssoCustomTokenClaims `mapstructure:"-"`
+}
+
+func (payload *ssoCustomTokenLoginPayload) Decode(data map[string]interface{}) skyerr.Error {
+	if err := mapstructure.Decode(data, payload); err != nil {
+		return skyerr.NewError(skyerr.BadRequest, "fails to decode the request payload")
+	}
+
+	if parsedToken, err := jwt.ParseWithClaims(
+		payload.TokenString,
+		&payload.Claims,
+		payload.keyFunc,
+	); err != nil {
+		return skyerr.NewError(skyerr.BadRequest, "fails to decode the request payload")
+	} else {
+		payload.Token = parsedToken
+	}
+
+	return payload.Validate()
+}
+
+func (payload *ssoCustomTokenLoginPayload) Validate() skyerr.Error {
+	claims := payload.Claims
+	if claims.Subject == "" {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: subject (sub) not specified",
+		)
+	}
+
+	if claims.ExpiresAt == 0 {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: expires at (exp) not specified",
+		)
+	}
+
+	if claims.IssuedAt == 0 {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: issued at (iat) not specified",
+		)
+	}
+
+	if claims.Valid() != nil {
+		return skyerr.NewError(
+			skyerr.InvalidCredentials,
+			"invalid token: token is not valid at this time",
+		)
+	}
+
+	return nil
+}
+
+/*
+SSOCustomTokenLoginHandler authenticate user with password
+
+The user can be either identified by username or password.
+
+curl -X POST -H "Content-Type: application/json" \
+  -d @- http://localhost:3000/ <<EOF
+{
+	"action": "sso:custom_token:login",
+	"token": "eyXXXXXXX",
+}
+EOF
+*/
+type SSOCustomTokenLoginHandler struct {
+	CustomTokenSecret string
+
+	TokenStore       authtoken.Store    `inject:"TokenStore"`
+	ProviderRegistry *provider.Registry `inject:"ProviderRegistry"`
+	HookRegistry     *hook.Registry     `inject:"HookRegistry"`
+	AssetStore       asset.Store        `inject:"AssetStore"`
+	AuthRecordKeys   [][]string         `inject:"AuthRecordKeys"`
+	AccessKey        router.Processor   `preprocessor:"accesskey"`
+	DBConn           router.Processor   `preprocessor:"dbconn"`
+	InjectPublicDB   router.Processor   `preprocessor:"inject_public_db"`
+	PluginReady      router.Processor   `preprocessor:"plugin_ready"`
+	preprocessors    []router.Processor
+}
+
+func (h *SSOCustomTokenLoginHandler) Setup() {
+	h.preprocessors = []router.Processor{
+		h.AccessKey,
+		h.DBConn,
+		h.InjectPublicDB,
+		h.PluginReady,
+	}
+}
+
+func (h *SSOCustomTokenLoginHandler) GetPreprocessors() []router.Processor {
+	return h.preprocessors
+}
+
+func (h *SSOCustomTokenLoginHandler) Handle(payload *router.Payload, response *router.Response) {
+	if h.CustomTokenSecret == "" {
+		response.Err = skyerr.NewError(
+			skyerr.NotConfigured,
+			"login with custom token requires CUSTOM_TOKEN_SECRET config",
+		)
+		return
+	}
```

Comment:
Logically I think this checking should happen at preprocessorr...?

Reply:
I am not sure about that... doing so we need to create a preprocessor to check if the custom token secret is configured, meaning that the preprocess have only one purpose and it is always used by this handler.

Okay?